### PR TITLE
[QoS] Verify the additional lossless queues and PGs in QoS test in dual ToR scenario

### PR DESCRIPTION
**Pull request by stephenxs**
Provide a WA to verify the additional lossless queues and PGs in the QoS test in the dual ToR scenario

1. Add a CLI option to specify whether it is to verify ports with or without additional lossless PGs and queues.
   It will collect all the dual ToR ports at the beginning of the test. If the option is on, the additional lossless PGs/queues will be verified in the corresponding tests
   The port information will always be collected regardless of whether the CLI option is fed because non-dual ToR ports must be used while the option is not fed, which means we need to understand the ports' role in any case.
   We use a Lua script to fetch such info because it is much faster than intensively using `sonic-db-cli`
2. Update the logic to fetch buffer profiles from BUFFER_QUEUE and BUFFER_PG table according to the CLI
3. Two additional profiles are introduced for verifying the additional lossless PGs/queues in XON and XOFF test if the CLI option is on
4. For the headroom pool test, all 4 DSCPs are passed to the PTF script as well as the CLI option. The additional DSCPs will be skipped by the PTF docker according to whether the ports are with additional lossless PGs/queues.
5. For the DSCP2queue mapping test, the CLI option is passed to the PTF script so that the latter is able to check the mapping according to the CLI option.

Signed-off-by: Stephen Sun <stephens@nvidia.com>

### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [x] 202012

### CODE REVIEW

**Comment by stephenxs** [comment]
FYI.
Using `sonic-db-cli` to fetch ports' information: ~55 seconds
```
2022-07-08 06:51:19 qos_sai_base.initDualTorPorts            L0092 INFO   | Starting fetching dual ToR info
2022-07-08 06:52:14 qos_sai_base.initDualTorPorts            L0123 INFO   | Finish fetching dual ToR info set([u'Ethernet96', u'Ethernet124', u'Ethernet92', u'Ethernet120', u'Ethernet68', u'Ethernet76', u'Ethernet108', u'Ethernet88', u'Ethernet72', u'Ethernet64', u'Ethernet116', u'Ethernet100', u'Ethernet80', u'Ethernet112', u'Ethernet104', u'Ethernet84'])
```
Using Lua script to do it: 1 second
```
2022-07-08 08:23:54 qos_sai_base.initDualTorPorts            L0130 INFO   | Starting fetching dual ToR info
2022-07-08 08:23:55 qos_sai_base.initDualTorPorts            L0138 INFO   | Finish fetching dual ToR info set([u'Ethernet96', u'Ethernet124', u'Ethernet92', u'Ethernet120', u'Ethernet68', u'Ethernet76', u'Ethernet108', u'Ethernet88', u'Ethernet72', u'Ethernet64', u'Ethernet116', u'Ethernet100', u'Ethernet80', u'Ethernet112', u'Ethernet104', u'Ethernet84'])
```

File: tests/qos/args/qos_sai_args.py [hunk 1]
```diff
@@ -46,3 +46,10 @@ def add_qos_sai_args(parser):
         help="QoS SAI comma separated list of source ports. Test currently expects exactly 1 source port",
     )
 
+    qos_group.addoption(
+        "--qos_dual_tor",
+        action="store",
+        type=str2bool,
+        default=False,
+        help="Test QoS on dual ToR ports"
+    )
```
Comment (bingwang-ms) [inline tests/qos/args/qos_sai_args.py:55]: Can you please clarify why do we need this option? Can we set `is_dual_tor = True` when `dualtor_ports` is not empty?
I have the ask because when we are running nightly tests with pipeline, we are normally using the same options for all testbeds.

File: tests/qos/qos_sai_base.py [hunk 2]
```diff
@@ -49,6 +52,60 @@ def isBufferInApplDb(self, dut_asic):
...
+            if neighbor ~= nil and string.sub(neighbor, -2, -1) == expected_neighbor_suffix then\
+                local peer_type = redis.call('HGET', 'DEVICE_NEIGHBOR_METADATA|' .. neighbor, 'type')\
+                if peer_type == expected_neighbor_type then\
+                    table.insert(result, string.sub(all_ports_with_neighbor[i], 17, -1))\
+                end\
+            end\
+        end\
+        return result\
+    "
+
```
Comment (bingwang-ms) [inline tests/qos/qos_sai_base.py:91]: The lua script is cool. Can you please move the `initDualTorPorts` method to some common place, such as `tests/common/dualtor/dual_tor_utils.py` ?

File: tests/qos/args/qos_sai_args.py (hunk 1 above)
Comment (stephenxs) [reply tests/qos/args/qos_sai_args.py:55]: Sure.
We need to test both scenarios (with/without additional lossless PG/Q) even if `dualtor_ports` is not empty, right?

File: tests/qos/qos_sai_base.py (hunk 2 above)
Comment (stephenxs) [reply tests/qos/qos_sai_base.py:91]: Sure.

File: tests/qos/args/qos_sai_args.py (hunk 1 above)
Comment (bingwang-ms) [reply tests/qos/args/qos_sai_args.py:55]: Yes, that makes sense. Thanks

File: tests/qos/qos_sai_base.py (hunk 2 above)
Comment (stephenxs) [reply tests/qos/qos_sai_base.py]: Fixed. Updated it to a fixture.

**Review by bingwang-ms — APPROVED** [review]

**Comment by neethajohn** [comment]
@stephenxs, can you please resolve the merge conflicts?

**Comment by stephenxs** [comment]
> @stephenxs, can you please resolve the merge conflicts?

Done.

**Comment by stephenxs** [comment]
> This pull request **introduces 1 alert** when merging [2fbb07c](https://github.com/Azure/sonic-mgmt/commit/2fbb07cba7affb4c4699a6c8e412b0fc82d57090) into [6319ace](https://github.com/Azure/sonic-mgmt/commit/6319acec81e2cca81bfacec92bb61ebc58b858e5) - [view on LGTM.com](https://lgtm.com/projects/g/Azure/sonic-mgmt/rev/pr-026d6ae4208de717defeda982f5757561b174e32)
> 
> **new alerts:**
> 
> * 1 for Unused import

This is a fake alarm.
```
from tests.common.dualtor.dual_tor_utils import dualtor_ports             # lgtm[py/unused-import]
  | NEWImport of 'dualtor_ports' is not used.
33 | 34
```
dualtor_ports is used as a fixture in `testParameter`

**Review by bingwang-ms — APPROVED** [review]

**Comment by stephenxs** [comment]
@wangxin can we cherry-pick this on to 202012?
